lib/eigensolve_quda.cpp

```diff
@@ -0,0 +1,948 @@
+                     residua[i]);
+        }
+      }
+
+      // Compute SVD if requested
+      time_svd = -clock();
+      if (eig_param->compute_svd) { computeSVD(kSpace, evals); }
+      time_svd += clock();
+    }
+
+    double total;
+
+    if (eig_param->compute_svd)
+      total = (time_e + time_mv + time_mb + time_svd) / CLOCKS_PER_SEC;
+    else
+      total = (time_e + time_mv + time_mb) / CLOCKS_PER_SEC;
+
+    if (getVerbosity() >= QUDA_SUMMARIZE) {
+      printfQuda("Time to solve problem using TRLM = %e\n", total);
+      printfQuda("Time spent using EIGEN           = %e  %.1f%%\n", time_e / CLOCKS_PER_SEC,
+                 100 * (time_e / CLOCKS_PER_SEC) / total);
+      printfQuda("Time spent in matVec             = %e  %.1f%%\n", time_mv / CLOCKS_PER_SEC,
+                 100 * (time_mv / CLOCKS_PER_SEC) / total);
+      printfQuda("Time spent in (multi)blas        = %e  %.1f%%\n", time_mb / CLOCKS_PER_SEC,
+                 100 * (time_mb / CLOCKS_PER_SEC) / total);
+      if (eig_param->compute_svd)
+        printfQuda("Time spent computing svd         = %e  %.1f%%\n", time_svd / CLOCKS_PER_SEC,
+                   100 * (time_svd / CLOCKS_PER_SEC) / total);
+    }
+    //---------------------------------------------------------------------------
+
+    // Local clean-up
+    delete r[0];
+
+    // Only save if outfile is defined
+    if (strcmp(eig_param->vec_outfile, "") != 0) {
+      if (getVerbosity() >= QUDA_SUMMARIZE) printfQuda("saving eigenvectors\n");
+      saveVectors(kSpace, eig_param->vec_outfile);
+    }
+
+    if (getVerbosity() >= QUDA_SUMMARIZE) {
+      printfQuda("*****************************\n");
+      printfQuda("***** END TRLM SOLUTION *****\n");
+      printfQuda("*****************************\n");
+    }
+  }
+
+  // Destructor
+  TRLM::~TRLM()
+  {
+    ritz_mat.clear();
+    ritz_mat.shrink_to_fit();
+    host_free(alpha);
+    host_free(beta);
+  }
+
+  // Thick Restart Member functions
+  //---------------------------------------------------------------------------
+  void TRLM::lanczosStep(std::vector<ColorSpinorField *> v, int j)
+  {
+
+    // Compute r = A * v_j - b_{j-i} * v_{j-1}
+    // r = A * v_j
+
+    time_ = -clock();
+
+    chebyOp(mat, *r[0], *v[j]);
+
+    // a_j = v_j^dag * r
+    alpha[j] = blas::reDotProduct(*v[j], *r[0]);
+
+    // r = r - a_j * v_j
+    blas::axpy(-alpha[j], *v[j], *r[0]);
+
+    int start = (j > num_keep) ? j - 1 : 0;
+    for (int i = start; i < j; i++) {
+
+      // r = r - b_{j-1} * v_{j-1}
+      blas::axpy(-beta[i], *v[i], *r[0]);
+    }
+
+    // Orthogonalise r against the Krylov space
+    if (j > 0)
+      for (int k = 0; k < 1; k++) blockOrthogonalize(v, r, j);
+
+    // b_j = ||r||
+    beta[j] = sqrt(blas::norm2(*r[0]));
+
+    // Prepare next step.
+    // v_{j+1} = r / b_j
+    blas::zero(*v[j + 1]);
+    blas::axpy(1.0 / beta[j], *r[0], *v[j + 1]);
+  }
+
+  void TRLM::reorder(std::vector<ColorSpinorField *> &kSpace)
```

Comment:
Not that I'd hold up the merge with this, but can we accomplish this with one of the standard C++ algorithms?

Reply:
What I think is needed here is to create a `std::pair` of eigenvalues and eigenvectors and then do `std::sort` on the eigenvalues.  As noted, not something to delay the pull on.

Reply:
Update: it looks like this has already been worked out in `lib/quda_arpack_interface.cpp`, let's grab it from there.